The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16864385104/job/47769259055) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47769259055) [(plain)](https://github.com/rust-lang/rust/commit/d2bfea290c341fc0d4ee71a270786e8147be2a53/checks/47769259055/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] askama test:false 0.186
error[E0308]: mismatched types
    --> src/librustdoc/passes/collect_intra_doc_links.rs:1029:16
     |
1029 |         && let Some(suffix) = ori_link.link.starts_with('!')
     |                ^^^^^^^^^^^^   ------------------------------ this expression has type `bool`
     |                |
     |                expected `bool`, found `Option<_>`
     |
     = note: expected type `bool`
```

</details>

--- a/src/librustdoc/passes/collect_intra_doc_links.rs
+++ b/src/librustdoc/passes/collect_intra_doc_links.rs
@@ -997,6 +997,7 @@ fn preprocess_link(
         }
     };
 
+    let is_shortcut_style = ori_link.kind == LinkType::ShortcutUnknown;
     // If there's no backticks, be lenient and revert to the old behavior.
     // This is to prevent churn by linting on stuff that isn't meant to be a link.
     // only shortcut links have simple enough syntax that they
@@ -1013,9 +1014,20 @@ fn preprocess_link(
     // | has backtick |    never ignore    |    never ignore   |
     // | no backtick  | ignore if url-like |    never ignore   |
     // |-------------------------------------------------------|
-    let ignore_urllike =
-        can_be_url || (ori_link.kind == LinkType::ShortcutUnknown && !ori_link.link.contains('`'));
+    let ignore_urllike = can_be_url || (is_shortcut_style && !ori_link.link.contains('`'));
     if ignore_urllike && should_ignore_link(path_str) {
+        return None;
+    }
+    // If we have an intra-doc link starting with `!` (which isn't `[!]` because this is the never type), we ignore it
+    // as it is never valid.
+    //
+    // The case is common enough because of cases like `#[doc = include_str!("../README.md")]` which often
+    // uses GitHub-flavored Markdown (GFM) admonitions, such as `[!NOTE]`.
+    if is_shortcut_style
+        && let Some(suffix) = ori_link.link.strip_prefix('!')
+        && !suffix.is_empty()
+        && suffix.chars().all(|c| c.is_ascii_alphabetic())
+    {
         return None;
     }
 

--- a/tests/rustdoc-ui/intra-doc/github-flavored-admonitions.rs
+++ b/tests/rustdoc-ui/intra-doc/github-flavored-admonitions.rs
@@ -1,0 +1,6 @@
+// regression test for https://github.com/rust-lang/rust/issues/141866
+//@ check-pass
+#![deny(rustdoc::broken_intra_doc_links)]
+
+//! > [!NOTE]
+//! > This should not cause any warnings

